library/common/extensions/filters/http/local_error/filter.cc

```diff
@@ -1,100 +1,28 @@
 #include "library/common/extensions/filters/http/local_error/filter.h"
 
-#include "envoy/http/codes.h"
 #include "envoy/server/filter_config.h"
 
-#include "source/common/grpc/common.h"
-#include "source/common/grpc/status.h"
-#include "source/common/http/codes.h"
-#include "source/common/http/header_map_impl.h"
-#include "source/common/http/utility.h"
-
-#include "library/common/http/headers.h"
-#include "library/common/types/c_types.h"
-
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace LocalError {
 
-namespace {
-
-// Envoy Mobile surfaces non-"OK" local responses as errors via callbacks rather than an HTTP
-// response. Here that response's "real" status code is mapped to an Envoy Mobile error code
-// which is passed through the response chain as a header. The status code is updated with
-// the sentinel value, 218 ("This is fine").
-void mapLocalHttpResponseToError(uint64_t status, Http::ResponseHeaderMap& headers) {
-  switch (status) {
-  case 408:
-    headers.addCopy(Http::InternalHeaders::get().ErrorCode, std::to_string(ENVOY_REQUEST_TIMEOUT));
-    break;
-  case 413:
-    headers.addCopy(Http::InternalHeaders::get().ErrorCode,
-                    std::to_string(ENVOY_BUFFER_LIMIT_EXCEEDED));
-    break;
-  case 503:
-    headers.addCopy(Http::InternalHeaders::get().ErrorCode,
-                    std::to_string(ENVOY_CONNECTION_FAILURE));
-    break;
-  default:
-    headers.addCopy(Http::InternalHeaders::get().ErrorCode, std::to_string(ENVOY_UNDEFINED_ERROR));
-  }
-
-  headers.setStatus(218);
-}
-
-} // namespace
-
-LocalErrorFilter::LocalErrorFilter() {}
-
-Http::FilterHeadersStatus LocalErrorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
-                                                          bool end_stream) {
-  // HACK: currently Envoy sends local replies in cases where an error ought to be
-  // surfaced via the error path. See https://github.com/lyft/envoy-mobile/issues/460
-
-  // Absence of EnvoyUpstreamServiceTime header implies this is a local reply, which we potentially
-  // map to a stream error.
-  if (!headers.get(Http::Headers::get().EnvoyUpstreamServiceTime).empty()) {
-    return Http::FilterHeadersStatus::Continue;
+Http::LocalErrorStatus LocalErrorFilter::onLocalReply(const LocalReplyData& reply) {
+  ENVOY_LOG(trace, "LocalErrorFilter::onLocalReply({}, {})", reply.code_, reply.details_);
+  ASSERT(decoder_callbacks_);
+  auto& info = decoder_callbacks_->streamInfo();
+  // TODO(goaway): set responseCode in upstream Envoy when responseCodDetails are set.
+  // ASSERT(static_cast<uint32_t>(reply.code_) == info.responseCode());
+  // TODO(goaway): follow up on the underscore discrepancy between these values.
+  // ASSERT(reply.details_ == info.responseCodeDetails().value());
+  info.setResponseCode(static_cast<uint32_t>(reply.code_));
```

Comment:
They almost match. One has some underscores subbed in for spaces, and only for certain error messages. They're still readable, but I'd like to resolve anyways.

Reply:
related to https://github.com/envoyproxy/envoy-mobile/pull/1592/files/bb8f73e01488c7f7f7dc3134ed626fa3cef54808#r694336361.

It looks like we have two sources for status code & details: stream info and the reply arg. Here it looks like we use the reply arg to a) modify stream info b) charge with the outlier detector. In my other thread below it looks like we directly use the reply arg and don't change anything about the stream info.

Reply:
Correct (also we don't have the reply arg available in the Http::Client). The only reason I'm changing stream info for the status code is because it isn't already set (which I assume is an oversight). The message is already set, and it's almost character-for-character identical, with the exception of the underscores.

It doesn't really matter a whole lot, because it's a string, and they're both readable. I think the correct place to reconcile the two would be in Envoy itself.

Reply:
@goaway that is interesting that the status code would not be set in response info. That seems like a bug to me.

Reply:
I guess if you'd like I would be fine relying on the reply arg in this PR and open an issue to delve into stream info's reliability.

Reply:
I can open a tracking issue.